### PR TITLE
[Refactor] Refactor HunyuanImage3 SigLIP2 ViT to vLLM layers

### PR DESCRIPTION
**Pull request by gcanlin**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose

- Rewrites the HunyuanImage3 SigLIP2 vision transformer from a HuggingFace-style padded implementation to vLLM-style packed sequence execution.
- Replaces separate Q/K/V projections with QKVParallelLinear and uses MMEncoderAttention for efficient encoder attention.
- Replaces MLP linear layers with TP-aware ColumnParallelLinear and RowParallelLinear.
- Packs real image patches using attention_mask, runs attention with cu_seqlens, then scatters outputs back to the padded batch shape.
- Adds weight-loading remapping for fused QKV parameters while skipping removed modules such as pooling outputs.
- Adds tensor-parallel/data-parallel aware execution paths for the ViT.

## Test Plan

```
vllm serve tencent/HunyuanImage-3.0-Instruct --omni   --port 8091   --stage-configs-path vllm_omni/model_executor/stage_configs/hunyuan_image3_i2t.yaml   --chat-template /tmp/hunyuan_image3_i2t_chat_template.jinja   --chat-template-content-format openai --mm-encoder-tp-mode data
```

```
vllm bench serve   --omni   --backend openai-chat-omni   --host 127.0.0.1 --port 8091   --endpoint /v1/chat/completions   --model tencent/HunyuanImage-3.0-Instruct   --dataset-name random-mm   --num-prompts 100   --request-rate inf   --max-concurrency 8   --random-input-len 256   --random-output-len 8   --random-range-ratio 0   --random-mm-base-items-per-request 1   --random-mm-num-mm-items-range-ratio 0   --random-mm-limit-mm-per-prompt '{"image":1,"video":0,"audio":0}'   --random-mm-bucket-config '{"(1024, 1024, 1)": 1.0}'   --extra-body '{"modalities":["text"]}'   --temperature 0 --ignore-eos   --percentile-metrics ttft   --metric-percentiles 50,90,99
```

## Benchmark Results

| Configuration | Request Throughput | Mean TTFT | P50 TTFT | P90 TTFT | P50 TPOT | Total Token Throughput | Observation |
|---|---:|---:|---:|---:|---:|---:|---|
| `max_num_batched_tokens=32768`, ViT-DP enabled | 1.11 req/s | 5575.87 ms | 6145.71 ms | 6917.02 ms | 138.79 ms | 5985.15 tok/s | Higher TTFT, best TPOT |
| `max_num_batched_tokens=32768`, ViT-DP disabled | 1.10 req/s | 5648.63 ms | 6240.56 ms | 6993.93 ms | 140.84 ms | 5913.41 tok/s | Slightly slower than ViT-DP |
| `max_num_batched_tokens=8192`, ViT-DP enabled | 1.13 req/s | 4387.20 ms | 4327.82 ms | 4341.27 ms | 375.44 ms | 6118.70 tok/s | Best TTFT, ViT-DP has little batch-level benefit |
| `max_num_batched_tokens=8192`, ViT-DP disabled | 1.13 req/s | 4413.86 ms | 4355.01 ms | 4374.77 ms | 377.15 ms | 6081.61 tok/s | Almost identical to ViT-DP enabled |

## Pairwise Comparison

| Comparison | Difference |
|---|---|
| `32768 + ViT-DP` vs `32768 + ViT-TP` | ViT-DP improves mean TTFT by about 73 ms and P50 TTFT by about 95 ms. The benefit exists, but is modest. |
| `8192 + ViT-DP` vs `8192 + ViT-TP` | ViT-DP improves mean TTFT by about 27 ms, which is close to benchmark noise. |
| `8192 + ViT-DP` vs `32768 + ViT-DP` | `8192` improves mean TTFT by about 1189 ms and P50 TTFT by about 1818 ms. |
| `32768 + ViT-DP` vs `8192 + ViT-DP` | `32768` improves P50 TPOT by about 237 ms, so decode-side amortization is better with the larger batch token budget. |

## Interpretation

The average input length is:

```text
539000 input tokens / 100 requests = 5390 input tokens per request
```

With `max_num_batched_tokens=8192`, a prefill step can usually fit only one request. With `max_num_batched_tokens=32768`, a prefill step can fit roughly six requests. This changes the scheduling behavior significantly.

`max_num_batched_tokens=8192` has much better TTFT because each prefill batch is smaller and finishes sooner. `max_num_batched_tokens=32768` has better TPOT because larger batches improve decode-side and scheduling amortization, but the larger prefill batch delays first-token latency.

ViT-DP helps only when there are enough real images in the same ViT forward pass to shard across TP ranks. In the `8192` case, the runtime log shows:

```text
[ViT-DP] images-in-forward=1 world_size=4 per_rank=1 padded_dummies=3
```

This means there is only one real image in the ViT batch. With TP size 4, the other three ranks receive dummy padded inputs, so ViT-DP has almost no real batch-parallel work to distribute.

## Conclusion

For this workload, `max_num_batched_tokens=8192` is better if the goal is lower TTFT. `max_num_batched_tokens=32768` is better if the goal is lower TPOT or better decode-side amortization.

The workload is long-input and short-output, so overall latency is dominated by TTFT. Under this workload, the effect of `max_num_batched_tokens` is much larger than the effect of ViT-DP. ViT-DP becomes more useful when `images-in-forward` approaches or exceeds the TP size.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by alex-jw-brooks** [comment]
Nice work! Is there a reason to keep the siglip2 implementation in vLLM Omni instead of using the siglip2navit implementation vLLM?

**Comment by gcanlin** [comment]
> Nice work! Is there a reason to keep the siglip2 implementation in vLLM Omni instead of using the siglip2navit implementation vLLM?

Thx! They don't seem to be the same. For examples, sliglip2 in hunyuan-image outputs the constant `max_num_patches` tokens but siglip2navit not. Correct me if I'm wrong.

**Comment by hsliuustc0106** [comment]
BLOCKER scan:
| Category            | Result                                  |
|---------------------|-----------------------------------------|
| Correctness         | PASS                                    |
| Reliability/Safety  | PASS                                    |
| Breaking Changes    | ISSUE (removed pooling head output)        |
| Test Coverage       | ISSUE (no e2e test results provided)     |
| Documentation       | ISSUE (no migration guide for removed head)|
| Security            | PASS                                    |

**BLOCKING:**

- **Breaking Changes** — This PR removes the pooling head (`Siglip2MultiheadAttentionPoolingHead`) from the SigLIP2 ViT, which changes the output from `BaseModelOutputWithPooling` to just `last_hidden_state`. If any external code relied on `pooler_output`, this will break. The `vision_model` call in `hunyuan_image3.py` was updated, but other consumers might exist.

- **Test Coverage** — The test plan shows a `vllm serve` command, but no actual test results are provided. No generated images, no latency/VRAM measurements, and no comparison with the old implementation. For a substantial refactor like this, we need evidence that the new implementation produces correct outputs.

- **Documentation** — The removal of the pooling head is a breaking change, but no migration guide or explanation is provided for external users who might be using `pooler_output`.

VERDICT: REQUEST_CHANGES

**Comment by gcanlin** [comment]
cc @Isotr0py @princepride

**Review by Gaohan123 — COMMENTED** [review]
Please post performance statistics before and after the PR. And add some image generation samples.

File: vllm_omni/model_executor/models/hunyuan_image3/hunyuan_image3.py [hunk 1]
```diff
@@ -1235,7 +1235,11 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
         self.time_embed = TimestepEmbedder(hidden_size=config.hidden_size)
 
         # vision
-        self.vision_model = Siglip2VisionTransformer(config.vit)
+        self.vision_model = Siglip2VisionTransformer(
```
Comment (Bounty-hunter) [inline vllm_omni/model_executor/models/hunyuan_image3/hunyuan_image3.py:1533]: Also need to change `pipeline_hunyuan_image3.py`, witch call `vllm_omni.model_executor.models.hunyuan_image3.siglip2:: Siglip2VisionTransformer`

**Review by Bounty-hunter — CHANGES_REQUESTED** [review]

**Comment by Bounty-hunter** [comment]
Run local test `tests/e2e/accuracy/test_hunyuan_image3.py`

**Comment by gcanlin** [comment]
~Hmmm. After enabling ViT DP, the performance is degraded. Perhaps it only benefit when requesting multi-image but we don't have this scenario.~

I didn't enable it successfully.

**Comment by gcanlin** [comment]
@Bounty-hunter @hsliuustc0106 @Gaohan123 I updated the latest benchmark result. PTAL

**Comment by Gaohan123** [comment]
Please post some comparison about accuracy before and after the PR. Thanks

File: vllm_omni/model_executor/models/hunyuan_image3/hunyuan_image3.py (hunk 1 above)
Comment (gcanlin) [reply vllm_omni/model_executor/models/hunyuan_image3/hunyuan_image3.py:1533]: The new added parameters is only for quantization. So we can keep the original calling in diffusion pipeline.

File: vllm_omni/model_executor/models/hunyuan_image3/siglip2.py [hunk 2]
```diff
@@ -65,470 +84,336 @@ def __init__(self, config):
...
-                Pixel values of shape (batch_size, max_num_patches, num_channels * patch_size * patch_size)
-            spatial_shapes (`List[Tuple[int, int]]`):
-                Spatial shapes of shape (batch_size, 2) to resize the positional embeddings to
-        """
+            pixel_values: Packed pixel values
+                (total_real_patches, num_channels * patch_size * patch_size)
+            spatial_shapes: Per-image spatial shapes (B, 2) as [(h, w), ...]
 
-        # Apply patch embeddings to already patchified pixel values
+        Returns:
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/hunyuan_image3/siglip2.py:100]: The pack uses `mask_bool` but `cu_seqlens` is computed from `spatial_shapes`. If these ever diverge (mask says N real tokens but h*w sums to M), the `output[mask_bool] = hidden_states` scatter will fail with a size mismatch or silently corrupt. Consider asserting `mask_bool.sum() == seq_lens.sum()` early — cheap insurance for debug time.

File: vllm_omni/model_executor/models/hunyuan_image3/siglip2.py [hunk 3]
```diff
@@ -65,470 +84,336 @@ def __init__(self, config):
...
         )
+        attn_output = attn_output.reshape(seq_length, self.num_heads_per_partition * self.head_dim)
 
-        attn_output = attn_output.transpose(1, 2).contiguous()
-        attn_output = attn_output.view(batch_size, q_len, self.embed_dim)
-
-        attn_output = self.out_proj(attn_output)
-
-        return attn_output, None
+        attn_output, _ = self.out_proj(attn_output)
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/hunyuan_image3/siglip2.py:205]: `num_heads_per_partition` is cached at init time via `is_vit_use_data_parallel()`. This works because DP mode is determined by config and doesn't change at runtime, but worth a brief note for future readers that this is intentionally static.

**Review by hsliuustc0106 — APPROVED** [review]
Clean refactor. Packed sequence processing, DP support, and weight loading all look correct. Benchmarks are comprehensive.

---

*Reviewed by Claude Code with glm-5.1*

**Comment by TaffyOfficial** [comment]
'''
Metric | Value | L20x Reference
-- | -- | --
COT similarity to reference | 0.9722 | 0.9644
COT prefix match | 29 | 29
Image-Image similarity | 92.5253 | 94.5538
SSIM | 0.2379 | 0.2420
PSNR (dB) | 13.73 | 14.10
'''

@gcanlin There is a drop in accuracy.

**Comment by gcanlin** [comment]
> '''
> 
> Metric	Value	L20x Reference
> COT similarity to reference	0.9722	0.9644
> COT prefix match	29	29
> Image-Image similarity	92.5253	94.5538
> SSIM	0.2379	0.2420
> PSNR (dB)	13.73	14.10
> '''		
> @gcanlin There is a drop in accuracy.

Could you help run again by using `mm_encoder_attn_backend: TORCH_SDPA` in stage 0?(BTW, would better to print this field to make sure it work. Sometimes the config doesn't take effect)

**Comment by TaffyOfficial** [comment]
COT similarity to reference: 0.9697  (ref 0.9644)
COT prefix match:            37      (ref 29)
Image-Image similarity:      92.6552 (ref 94.5538)
SSIM:                        0.2371  (ref 0.2420)
PSNR (dB):                   14.02   (ref 14.10)
